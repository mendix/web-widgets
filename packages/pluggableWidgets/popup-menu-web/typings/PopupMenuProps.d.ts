/**
 * This file was generated from PopupMenu.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { ComponentType, CSSProperties, ReactNode } from "react";
import { ActionValue, DynamicValue } from "mendix";

export type ItemTypeEnum = "item" | "divider";

export type StyleClassEnum = "defaultStyle" | "inverseStyle" | "primaryStyle" | "infoStyle" | "successStyle" | "warningStyle" | "dangerStyle";

export interface BasicItemsType {
    itemType: ItemTypeEnum;
    caption?: DynamicValue<string>;
    visible?: DynamicValue<boolean>;
    action?: ActionValue;
    styleClass: StyleClassEnum;
}

export interface CustomItemsType {
    content: ReactNode;
    visible?: DynamicValue<boolean>;
    action?: ActionValue;
}

export type TriggerEnum = "onclick" | "onhover";

export type HoverCloseOnEnum = "onClickOutside" | "onHoverLeave";

export type PositionEnum = "left" | "right" | "top" | "bottom";

export type ClippingStrategyEnum = "absolute" | "fixed";

export interface BasicItemsPreviewType {
    itemType: ItemTypeEnum;
    caption: string;
    visible: string;
    action: {} | null;
    styleClass: StyleClassEnum;
}

export interface CustomItemsPreviewType {
    content: { widgetCount: number; renderer: ComponentType<{ children: ReactNode; caption?: string }> };
    visible: string;
    action: {} | null;
}

export interface PopupMenuContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex?: number;
    advancedMode: boolean;
    menuTrigger: ReactNode;
    basicItems: BasicItemsType[];
    customItems: CustomItemsType[];
    trigger: TriggerEnum;
    hoverCloseOn: HoverCloseOnEnum;
    position: PositionEnum;
    clippingStrategy: ClippingStrategyEnum;
    menuToggle: boolean;
}

export interface PopupMenuPreviewProps {
    /**
     * @deprecated Deprecated since version 9.18.0. Please use class property instead.
     */
    className: string;
    class: string;
    style: string;
    styleObject?: CSSProperties;
    readOnly: boolean;
    renderMode?: "design" | "xray" | "structure";
    advancedMode: boolean;
    menuTrigger: { widgetCount: number; renderer: ComponentType<{ children: ReactNode; caption?: string }> };
    basicItems: BasicItemsPreviewType[];
    customItems: CustomItemsPreviewType[];
    trigger: TriggerEnum;
    hoverCloseOn: HoverCloseOnEnum;
    position: PositionEnum;
    clippingStrategy: ClippingStrategyEnum;
    menuToggle: boolean;
}
