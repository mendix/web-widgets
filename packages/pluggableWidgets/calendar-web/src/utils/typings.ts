import { ObjectItem } from "mendix";
import { CalendarProps as ReactCalendarProps } from "react-big-calendar";
import { withDragAndDropProps } from "react-big-calendar/lib/addons/dragAndDrop";

export interface CalendarEvent {
    title: string;
    start: Date;
    end: Date;
    allDay: boolean;
    color?: string;
    item: ObjectItem;
}

export type EventDropOrResize = {
    event: CalendarEvent;
    start: Date;
    end: Date;
};

export interface DragAndDropCalendarProps<TEvent extends object = Event, TResource extends object = object>
    extends ReactCalendarProps<TEvent, TResource>,
        withDragAndDropProps<TEvent, TResource> {}
