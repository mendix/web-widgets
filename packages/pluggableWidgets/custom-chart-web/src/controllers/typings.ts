import { ActionValue, EditableValue } from "mendix";

export interface ControllerProps {
    configurationOptions: string;
    dataAttribute?: EditableValue<string>;
    dataStatic: string;
    eventDataAttribute?: EditableValue<string>;
    layoutAttribute?: EditableValue<string>;
    layoutStatic: string;
    onClick?: ActionValue;
    sampleData: string;
    sampleLayout: string;
}
