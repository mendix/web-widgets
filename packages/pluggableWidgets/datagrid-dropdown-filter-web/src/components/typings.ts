import { RefFilterStore } from "@mendix/widget-plugin-dropdown-filter/stores/RefFilterStore";
import { StaticSelectFilterStore } from "@mendix/widget-plugin-dropdown-filter/stores/StaticSelectFilterStore";

export interface EnumFilterAPI {
    filterStore: StaticSelectFilterStore;
    parentChannelName?: string;
}

export interface RefFilterAPI {
    filterStore: RefFilterStore;
    parentChannelName?: string;
}
