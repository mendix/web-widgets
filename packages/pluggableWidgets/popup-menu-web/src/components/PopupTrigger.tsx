import { useMergeRefs } from "@floating-ui/react";
import { createElement, forwardRef, PropsWithChildren, ReactElement, RefObject } from "react";
import { usePopupContext } from "../hooks/usePopupContext";

export const PopupTrigger = forwardRef(
    ({ children }: PropsWithChildren, propRef: RefObject<HTMLDivElement>): ReactElement => {
        const { getReferenceProps, open, refs } = usePopupContext();
        const childrenRef = (children as any).ref;
        const ref = useMergeRefs([refs.setReference, propRef, childrenRef]);

        return (
            <div
                className={"popupmenu-trigger"}
                ref={ref}
                data-state={open ? "open" : "closed"}
                {...getReferenceProps?.({
                    onClick: e => {
                        e.preventDefault();
                        e.stopPropagation();
                    }
                })}
            >
                {children}
            </div>
        );
    }
);
