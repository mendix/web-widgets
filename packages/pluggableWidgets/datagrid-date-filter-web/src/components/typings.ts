import { Date_InputFilterInterface } from "@mendix/widget-plugin-filtering/typings/InputFilterInterface";

export interface DateFilterProps {
    filterStore: Date_InputFilterInterface;
    parentChannelName?: string;
}
