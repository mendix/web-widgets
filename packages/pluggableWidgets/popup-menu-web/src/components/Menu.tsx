import classNames from "classnames";
import { ActionValue } from "mendix";
import { createElement, ReactElement, useEffect, useRef } from "react";
import { createPortal } from "react-dom";
import { BasicItemsType, CustomItemsType, PopupMenuContainerProps, PositionEnum } from "../../typings/PopupMenuProps";
import { useOnClickOutside } from "@mendix/widget-kit-web/hooks/useOnClickOutside";
import { useMenuPlacement } from "../utils/useMenuPlacement";

import {
    isBehindElement,
    isBehindRandomElement,
    isElementPartiallyOffScreen,
    isElementVisibleByUser,
    moveAbsoluteElementOnScreen,
    unBlockAbsoluteElementBottom,
    unBlockAbsoluteElementLeft,
    unBlockAbsoluteElementRight,
    unBlockAbsoluteElementTop
} from "../utils/document";

export interface MenuProps extends PopupMenuContainerProps {
    anchorElement: HTMLDivElement;
    onCloseRequest: () => void;
    onItemClick: (itemAction: ActionValue) => void;
}
export function Menu(props: MenuProps): ReactElement {
    const popupRef = useRef<HTMLUListElement>(null);
    const anchorElement = props.anchorElement;
    const popupStyles = useMenuPlacement(anchorElement, props.position);

    useOnClickOutside(popupRef, props.onCloseRequest);

    useEffect(() => {
        if (popupRef.current) {
            correctPosition(popupRef.current, props.position);
        }
    }, [popupRef.current, props.position]);

    const menuOptions = createMenuOptions(props, props.onItemClick);

    return createPortal(
        <div className="widget-popupmenu-root">
            <ul
                ref={popupRef}
                style={popupStyles}
                className={classNames(
                    "popupmenu-menu",
                    `popupmenu-position-${props.position}`,
                    "popup-portal",
                    props.class
                )}
            >
                {menuOptions}
            </ul>
        </div>,
        anchorElement
    );
}

function checkVisibility(item: BasicItemsType | CustomItemsType): boolean {
    if (Object.prototype.hasOwnProperty.call(item, "visible")) {
        return !!item.visible?.value;
    }
    return true;
}

function createMenuOptions(
    props: PopupMenuContainerProps,
    handleOnClickItem: (itemAction?: ActionValue) => void
): ReactElement[] {
    if (!props.advancedMode) {
        return props.basicItems
            .filter(item => checkVisibility(item))
            .map((item, index) => {
                if (item.itemType === "divider") {
                    return <li key={index} className={"popupmenu-basic-divider"} />;
                } else {
                    const pickedStyle =
                        item.styleClass !== "defaultStyle"
                            ? "popupmenu-basic-item-" + item.styleClass.replace("Style", "")
                            : "";
                    return (
                        <li
                            key={index}
                            className={classNames("popupmenu-basic-item", pickedStyle)}
                            onClick={e => {
                                e.preventDefault();
                                e.stopPropagation();
                                handleOnClickItem(item.action);
                            }}
                        >
                            {item.caption?.value ?? ""}
                        </li>
                    );
                }
            });
    } else {
        return props.customItems
            .filter(item => checkVisibility(item))
            .map((item, index) => (
                <li
                    key={index}
                    className={"popupmenu-custom-item"}
                    onClick={e => {
                        e.preventDefault();
                        e.stopPropagation();
                        handleOnClickItem(item.action);
                    }}
                >
                    {item.content}
                </li>
            ));
    }
}

function correctPosition(element: HTMLElement, position: PositionEnum): void {
    const dynamicDocument: Document = element.ownerDocument;
    const dynamicWindow = dynamicDocument.defaultView as Window;
    let boundingRect: DOMRect = element.getBoundingClientRect();
    const isOffScreen = isElementPartiallyOffScreen(dynamicWindow, boundingRect);
    if (isOffScreen) {
        moveAbsoluteElementOnScreen(dynamicWindow, element, boundingRect);
    }

    boundingRect = element.getBoundingClientRect();
    const blockingElement = isBehindRandomElement(dynamicDocument, element, boundingRect, 3, "popupmenu");
    if (blockingElement && isElementVisibleByUser(dynamicDocument, dynamicWindow, blockingElement)) {
        unBlockAbsoluteElement(element, boundingRect, blockingElement.getBoundingClientRect(), position);
    } else if (blockingElement) {
        let node = blockingElement;
        do {
            if (isBehindElement(element, node, 3) && isElementVisibleByUser(dynamicDocument, dynamicWindow, node)) {
                return unBlockAbsoluteElement(element, boundingRect, node.getBoundingClientRect(), position);
            } else if (node.parentElement) {
                node = node.parentElement as HTMLElement;
            } else {
                break;
            }
        } while (node.parentElement);
    }
}

function unBlockAbsoluteElement(
    element: HTMLElement,
    boundingRect: DOMRect,
    blockingElementRect: DOMRect,
    position: PositionEnum
): void {
    switch (position) {
        case "left":
            unBlockAbsoluteElementLeft(element, boundingRect, blockingElementRect);
            unBlockAbsoluteElementBottom(element, boundingRect, blockingElementRect);
            break;
        case "right":
            unBlockAbsoluteElementRight(element, boundingRect, blockingElementRect);
            unBlockAbsoluteElementBottom(element, boundingRect, blockingElementRect);
            break;
        case "top":
            unBlockAbsoluteElementTop(element, boundingRect, blockingElementRect);
            break;
        case "bottom":
            unBlockAbsoluteElementBottom(element, boundingRect, blockingElementRect);
            break;
    }
}
