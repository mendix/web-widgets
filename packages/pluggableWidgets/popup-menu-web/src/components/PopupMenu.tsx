import { executeAction } from "@mendix/widget-plugin-platform/framework/execute-action";
import classNames from "classnames";
import { ActionValue } from "mendix";
import { createElement, ReactElement, useCallback, useEffect, useState } from "react";
import { PopupMenuContainerProps } from "../../typings/PopupMenuProps";
import { usePopup } from "../hooks/usePopup";
import { Menu } from "./Menu";
import { PopupContext } from "./PopupContext";
import { PopupTrigger } from "./PopupTrigger";

export interface PopupMenuProps extends PopupMenuContainerProps {
    preview?: boolean;
}

export function PopupMenu(props: PopupMenuProps): ReactElement {
    const preview = !!props.preview;
    const [visibility, setVisibility] = useState(preview && props.menuToggle);
    const open = visibility;
    const popup = usePopup({
        open,
        onOpenChange: setVisibility,
        placement: props.position,
        trigger: props.trigger,
        clippingStrategy: props.clippingStrategy
    });

    const handleOnClickItem = useCallback((itemAction?: ActionValue): void => {
        setVisibility(false);
        executeAction(itemAction);
    }, []);

    useEffect(() => {
        setVisibility(props.menuToggle);
    }, [props.menuToggle]);

    return (
        <PopupContext.Provider value={popup}>
            <div className={classNames("popupmenu", props.class)}>
                <PopupTrigger>{props.menuTrigger}</PopupTrigger>
                <Menu {...props} onItemClick={handleOnClickItem} />
            </div>
        </PopupContext.Provider>
    );
}
