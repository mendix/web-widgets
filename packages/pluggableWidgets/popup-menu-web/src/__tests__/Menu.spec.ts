import { shallow, ShallowWrapper } from "enzyme";
import { createElement } from "react";
import { BasicItemsType, CustomItemsType } from "../../typings/PopupMenuProps";
import { Menu, MenuProps } from "../components/Menu";
import { actionValue, dynamicValue } from "@mendix/pluggable-widgets-commons";
import { ValueStatus } from "mendix";

jest.useFakeTimers();

const positionObserverValues = { top: 10, bottom: 20, left: 30, height: 1, width: 2 };
const usePositionMock = jest.fn(() => positionObserverValues);
jest.mock("@mendix/pluggable-widgets-commons/dist/components/web", () => ({
    usePositionObserver: () => usePositionMock()
}));

describe("Popup menu", () => {
    const createPopupMenu = (props: MenuProps): ShallowWrapper<MenuProps, {}> => shallow(createElement(Menu, props));
    const basicItemProps: BasicItemsType = {
        itemType: "item",
        caption: dynamicValue("Caption"),
        styleClass: "defaultStyle"
    };
    const customItemProps: CustomItemsType = { content: createElement("div", null, null) };

    const defaultProps: MenuProps = {
        name: "popup-menu",
        class: "mx-popup-menu",
        tabIndex: -1,
        trigger: "onclick",
        menuToggle: false,
        menuTrigger: createElement("button", null, "Trigger"),
        advancedMode: false,
        position: "bottom",
        basicItems: [
            basicItemProps,
            { itemType: "divider", caption: dynamicValue("Caption"), styleClass: "defaultStyle" }
        ],
        customItems: [customItemProps],
        anchorElement: document.createElement("div"),
        onItemClick: jest.fn(),
        onCloseRequest: jest.fn()
    };

    it("renders popup menu", () => {
        const popupMenu = createPopupMenu(defaultProps);
        popupMenu.update();

        expect(popupMenu).toMatchSnapshot();
    });

    describe("with basic items", () => {
        it("renders", () => {
            const popupMenu = createPopupMenu(defaultProps);
            expect(popupMenu.find(".popupmenu-basic-item")).toHaveLength(1);
        });

        it("triggers action", () => {
            basicItemProps.action = actionValue();
            const popupMenu = createPopupMenu(defaultProps);
            const preventDefaultAction = jest.fn();
            const stopPropagationAction = jest.fn();
            const event: any = {
                preventDefault: preventDefaultAction,
                stopPropagation: stopPropagationAction,
                target: {}
            };
            popupMenu.find(".popupmenu-basic-item").prop("onClick")!(event);

            expect(stopPropagationAction).toHaveBeenCalled();
            expect(preventDefaultAction).toHaveBeenCalled();
            expect(defaultProps.onItemClick).toHaveBeenCalledTimes(1);
        });

        it("renders basic items without hidden items", () => {
            const basicItem: BasicItemsType = {
                ...basicItemProps,
                visible: {
                    value: false,
                    status: ValueStatus.Available
                }
            };

            const popupMenu = createPopupMenu({
                ...defaultProps,
                basicItems: [
                    basicItem,
                    { itemType: "divider", caption: dynamicValue("Caption"), styleClass: "defaultStyle" }
                ]
            });
            expect(popupMenu.find(".popupmenu-basic-item")).toHaveLength(0);
        });

        it("renders with style Inverse", () => {
            basicItemProps.styleClass = "inverseStyle";
            const popupMenu = createPopupMenu(defaultProps);

            expect(popupMenu.find(".popupmenu-basic-item-inverse")).toHaveLength(1);
        });
        it("renders with style Primary", () => {
            basicItemProps.styleClass = "primaryStyle";
            const popupMenu = createPopupMenu(defaultProps);

            expect(popupMenu.find(".popupmenu-basic-item-primary")).toHaveLength(1);
        });
        it("renders with style Info", () => {
            basicItemProps.styleClass = "infoStyle";
            const popupMenu = createPopupMenu(defaultProps);

            expect(popupMenu.find(".popupmenu-basic-item-info")).toHaveLength(1);
        });
        it("renders with style Success", () => {
            basicItemProps.styleClass = "successStyle";
            const popupMenu = createPopupMenu(defaultProps);

            expect(popupMenu.find(".popupmenu-basic-item-success")).toHaveLength(1);
        });
        it("renders with style Warning", () => {
            basicItemProps.styleClass = "warningStyle";
            const popupMenu = createPopupMenu(defaultProps);

            expect(popupMenu.find(".popupmenu-basic-item-warning")).toHaveLength(1);
        });
        it("renders with style Danger", () => {
            basicItemProps.styleClass = "dangerStyle";
            const popupMenu = createPopupMenu(defaultProps);

            expect(popupMenu.find(".popupmenu-basic-item-danger")).toHaveLength(1);
        });
    });

    describe("with custom items", () => {
        beforeEach(() => {
            defaultProps.advancedMode = true;
        });

        it("renders", () => {
            const popupMenu = createPopupMenu(defaultProps);

            expect(popupMenu.find(".popupmenu-custom-item")).toHaveLength(1);
        });

        it("renders custom items without hidden items", () => {
            const customItem: CustomItemsType = {
                ...customItemProps,
                visible: {
                    value: false,
                    status: ValueStatus.Available
                }
            };
            const popupMenu = createPopupMenu({
                ...defaultProps,
                customItems: [customItem]
            });
            expect(popupMenu.find(".popupmenu-custom-item")).toHaveLength(0);
        });

        it("triggers action", () => {
            const popupMenu = createPopupMenu(defaultProps);
            const onClick = popupMenu.find(".popupmenu-custom-item").prop("onClick");
            if (onClick) {
                const preventDefaultAction = jest.fn();
                const event: any = {
                    preventDefault: preventDefaultAction,
                    target: {}
                };
                onClick(event);
                expect(preventDefaultAction).toHaveBeenCalled();
                expect(defaultProps.onItemClick).toHaveBeenCalledTimes(1);
            }
        });
    });
});
