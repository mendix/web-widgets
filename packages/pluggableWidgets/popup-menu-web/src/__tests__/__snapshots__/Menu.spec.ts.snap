// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup menu renders popup menu 1`] = `
<Portal
  containerInfo={<div />}
>
  <div
    className="widget-popupmenu-root"
  >
    <ul
      className="popupmenu-menu popupmenu-position-bottom popup-portal mx-popup-menu"
      style={
        {
          "bottom": "initial",
          "display": "flex",
          "left": 0,
          "position": "absolute",
          "top": 1,
          "transform": "none",
          "zIndex": 1,
        }
      }
    >
      <li
        className="popupmenu-basic-item"
        key="0"
        onClick={[Function]}
      >
        Caption
      </li>
      <li
        className="popupmenu-basic-divider"
        key="1"
      />
    </ul>
  </div>
</Portal>
`;
